Faster guid and other xml speedups

christopherlam opened the pull request:
this does shave off ~5-10%~ 20% xml loading time.

jralls commented on libgnucash/app-utils/gnc-autoclear.cpp:
```diff
@@ -79,7 +79,7 @@ using SplitVec = std::vector<Split*>;
 struct Solution
 {
-    bool abort = false;
+    std::optional<std::string> abort;
```
Make `abort` a `const char *` and the two strings `static const char *` members. Saves the string alloc and converting char* -> string and back.

christopherlam commented on libgnucash/backend/xml/sixtp-dom-parsers.cpp:
```diff
@@ -705,77 +707,62 @@ dom_tree_to_commodity_ref_no_engine (xmlNodePtr node, QofBook* book)
-                                       gnc_commodity_get_namespace (daref),
-                                       gnc_commodity_get_mnemonic (daref));
-
-    gnc_commodity_destroy (daref);
```
This last commit is a good one. Every transaction has a currency node, which would be creating a new gnc_commodity and immediately destroying it. Refactor will avoid this wasteful trip.

jralls commented on libgnucash/backend/xml/sixtp-dom-parsers.cpp:
```diff
@@ -92,6 +98,10 @@ dom_tree_to_guid (xmlNodePtr node)
+    if (auto txt = dom_node_to_text (node))
+        if (gint64 daint; string_to_gint64 (txt, &daint))
+            return new KvpValue{daint};
+
```
This is sloppy. At least put the new code after the decls and use `text` instead of creating a new `txt` and `daint`. Better yet would be 
```
 {
    gchar* text{dom_node_to_text (node)};
    gint64 daint;
    gboolean valid_int{FALSE};

    if (text)
        valid_int = string_to_gint64 (text, &daint);
    else
    {
        text = dom_tree_to_text (node);
        valid_int = string_to_gint64 (text, &daint);
        g_free (text);
    }
    if (valid_int)
        return new KvpValue {daint};
    return nullptr
}
```

jralls replied:
Note you can make it even cleaner by changing `string_to_gint64` and friends to take a `const std::string_view` instead of a `char*` and `dom_tree_to_text` to return a `std::string`. Then you can write
```
{
     auto text{dom_node_to_text(node)};
     if (!text)
         text = dom_tree_to_text (node);
     gint64 daint;
     if (string_to_gint64 (text, &daint))
         return new KvpValue {daint};
     return nullptr;
}
```

christopherlam replied:
> Note you can make it even cleaner by changing `string_to_gint64` and friends to take a `const std::string_view` instead of a `char*` and `dom_tree_to_text` to return a `std::string`. Then you can write
> 
> ```
> {
>      auto text{dom_node_to_text(node)};
>      if (!text)
>          text = dom_tree_to_text (node);
>      gint64 daint;
>      if (string_to_gint64 (text, &daint))
>          return new KvpValue {daint};
>      return nullptr;
> }
> ```

hmm then `text` will be a `const char*` or `std::string`?

jralls replied:
> hmm then text will be a const char* or std::string?

Oh, right. It would have to be 
```
     std::string tree_text{};
     std::string_view text{dom_node_to_text(node)};
     if (text.empty())
     {
          tree_text = dom_tree_to_text(node);
          text = tree_text;
      }
      ...
```

christopherlam replied:
> > hmm then text will be a const char* or std::string?
> 
> Oh, right. It would have to be
> 
> ```
>      std::string tree_text{};
>      std::string_view text{dom_node_to_text(node)};
>      if (text.empty())
>      {
>           tree_text = dom_tree_to_text(node);
>           text = tree_text;
>       }
>       ...
> ```

Not quite because the dom_node_to_text(node) needs to be freed

jralls replied:
> Not quite because the dom_node_to_text(node) needs to be freed

No it doesn't, it's `reinterpret_cast<const char*>(node->children->content)`. Isn't that why you have separate a separate block with no free for it?

christopherlam replied:
> > Not quite because the dom_node_to_text(node) needs to be freed
> 
> No it doesn't, it's `reinterpret_cast<const char*>(node->children->content)`. Isn't that why you have separate a separate block with no free for it?

oops my bad... I mean `dom_tree_to_text` needs to be `g_free`d.

jralls replied:
It does now, that's why I said

> Note you can make it even cleaner by changing string_to_gint64 and friends to take a const std::string_view instead of a char* and dom_tree_to_text to return a std::string.

christopherlam replied:
> It does now, that's why I said
> 
> > Note you can make it even cleaner by changing string_to_gint64 and friends to take a const std::string_view instead of a char*

this is easily done

>> and dom_tree_to_text to return a std::string.

and this one less so because it can return `nullptr` for invalid conditions, and is heavily used by `backend/xml`. should it return `optional<string>`?

jralls replied:
That's the current preferred way, and you do have some practice with it. ;-)

christopherlam replied:
> That's the current preferred way, and you do have some practice with it. ;-)

Ok this is now present in last commit ... It's technically another layer of copying from xmlChar to char* to `optional<string>` to the engine data though. We can build the string directly by walking the XML tree though.

christopherlam replied:
> > That's the current preferred way, and you do have some practice with it. ;-)
> 
> Ok this is now present in last commit ... It's technically another layer of copying from xmlChar to char* to `optional<string>` to the engine data though. We can build the string directly by walking the XML tree though.

I've reverted the `dom_tree_to_text` change; it does not offer any speedup. The various engine setters (eg entry + kvp) using `dom_node_to_text` seem more effective; it reduces the unwanted temporary char* from 21000 to 350.

jralls replied:
It's not meant to be a speedup, it's to allow cleaner code and safer memory management. As long as it isn't a significant slowdown (and it shouldn't be, see below) it's still better.

> It's technically another layer of copying from xmlChar to char* to optional<string>

It's not. `xmlNodeListGetString` allocates a string and `temp` points at it. The old code allocates a copy with `strdup` and the new code allocates a copy in the `std::string`. No change. It's [theoretically possible](https://en.cppreference.com/w/cpp/string/basic_string/basic_string) to make a std::string class that uses libxml2's allocator and could use the xmlChar* directly from `xmlNodeListGetString`, eliminating the second allocation but I don't think we want to go there.

Note that `return rv` move-constructs the std::optional from rv into the caller's stack frame, see the example in [cpp_reference](https://en.cppreference.com/w/cpp/utility/optional/optional.html).

jralls commented on libgnucash/engine/guid.cpp:
```diff
@@ -266,15 +256,9 @@ guid_hash_to_guint (gconstpointer ptr)
-    return hash;
+    const GncGUID* g = static_cast<const GncGUID*>(ptr);
+    const std::string_view data{reinterpret_cast<const char*>(g->reserved), GUID_DATA_SIZE};
+    return std::hash<std::string_view>{}(data);
```
`std::hash` produces a `size_t` that fits in an `gint` only on 32-bit systems. You need to change the return type here and in all uses.

jralls replied:
Also it would be better to make a gnc:GUID member function and wrap it for use in gnc-sx-instance-model.c and call it directly in import-pending-matches.cpp.

christopherlam replied:
> `std::hash` produces a `size_t` that fits in an `gint` only on 32-bit systems. You need to change the return type here and in all uses.

as follows?

```cpp

guint
guid_hash_to_guint (gconstpointer ptr)
{
    if (!ptr)
    {
        PERR ("received nullptr guid pointer.");
        return 0;
    }
    const GncGUID* g = static_cast<const GncGUID*>(ptr);
    const std::string_view data{reinterpret_cast<const char*>(g->reserved), GUID_DATA_SIZE};
    uint64_t h = std::hash<std::string_view>{}(data);
    return static_cast<guint>(h ^ (h >> 32));
}
```

jralls replied:
`h ^ (h >> 32)` is a weird thing to do:
```
>>> hash=6919333181322027406
>>> hex(hash)
'0x600668de4345e18e'
>>> hex(hash >> 32)
'0x600668de'
>>> hex(hash ^ (hash >> 32))
'0x600668de23438950'
```
`6919333181322027406` comes from `[std::hash](https://en.cppreference.com/w/cpp/utility/hash.html)<[std::string_view](https://en.cppreference.com/w/cpp/string/basic_string_view.html)>{}("A"sv)`

You might as well just chop off the top half.

I looked again at the original hash func,
```
guint hash {0};
    std::for_each (temp.begin (), temp.end (), [&hash] (unsigned char a) {
        hash <<=4;
        hash |= a;
    });
```
It also shifts the top half of the GUID to oblivion. 

A GUID is already a random number so simply returning `*(std::reinterpret_cast<guint*>(&g->reserved[12]))` should be a good enough hash for our purposes.

jralls reviewed (changes requested):
Nice optimizations.

jralls commented:
Heh. The failing test is the one that #2158 fixes. Arch must have picked up guile 3.0.11.

christopherlam committed:
[backend/xml] dom_tree_to_text returns std::optional<std::string>

jralls commented on libgnucash/backend/xml/gnc-transaction-xml-v2.cpp:
```diff
@@ -193,6 +193,18 @@ gnc_transaction_dom_tree_create (Transaction* trn)
+    if (const char* txt = dom_node_to_text(node)) func(obj, txt);
+    else if (gchar* txt = dom_tree_to_text(node)) { func (obj, txt); g_free (txt); }
+    else return FALSE;
+    return TRUE;
```
A little more white space please.

jralls commented on libgnucash/backend/xml/gnc-budget-xml-v2.cpp:
```diff
@@ -89,11 +89,10 @@ static inline gboolean
 
-    func (bgt, txt);
-    g_free (txt);
+    func (bgt, txt->c_str());
```
Looks like this and bill-term above are expecting a std::string* from `dom_tree_to_text`.

jralls replied:
Oh, never mind, I missed that this is the commit that adds that.

jralls commented on libgnucash/backend/xml/gnc-entry-xml-v2.cpp:
```diff
@@ -401,44 +399,41 @@ entry_idiscount_handler (xmlNodePtr node, gpointer entry_pdata)
-    return ret;
+    auto entry = pdata->entry;
+    return parse_and_set (node, [entry](auto str)
+    {
```
This takes close reading to notice that the braces are the start of the lambda. Maybe declare the lambda first, then call:
```
auto entry = pdata->entry;
auto set_from_entry = 
    [entry](auto str)
    { 
          //...
    };
return parse_and_set (node, set_from_entry);
```

jralls reviewed (changes requested):
Please follow the coding guidelines!
```
if (foo)
    bar();
``` 
not `if (foo) bar();` and especially not `if (foo) { bar(); baz(); }`

jralls commented on libgnucash/backend/xml/sixtp-dom-parsers.cpp:
```diff
@@ -60,6 +60,11 @@ dom_tree_to_guid (xmlNodePtr node)
         return {};
     }
 
+    if (auto type_str = dom_node_to_text ((xmlNodePtr)node->properties); !strcmp (type_str, "guid"))
```
Better use `g_strcmp0` here, `type_str` could be NULL and `strcmp` isn't NULL-safe.

christopherlam committed:
[sixtp-dom-parsers.h] add templated functions for node text

* apply_xmlnode_text (node, f, optional default_val)
  retrieves text; and calls and returns f(text)
  on failure (i.e. node==null || node has no text), it returns default_val

* apply_xmlnode_text (node, obj, f)
  retrieves text; and calls and returns f(obj, text)
  on success, it will return true, on failure it returns false

christopherlam committed:
[backend/xml] use apply_xmlnode_text

christopherlam commented:
Any thoughts about naming the two templates functions in the header?

jralls commented:
How about `apply_string` and `apply_xml_node_text`? I'm not keen on starting either name with `gnc` because the function isn't exported outside of the xml backend and `xml` doesn't really belong up front: Neither does anything to the DOM tree and the first one doesn't even have involve xml.

christopherlam commented:
Ok how about apply_node_text and apply_node_text_to_object

jralls commented:
> the first one doesn't even have involve xml

Oops, that's wrong, it's the one getting the text from the xmlNode. So `apply_string` is wrong. `apply_node_text` isn't specific enough, it should be `apply_xml_node_text` (or `apply_xmlnode_text` since the first arg is an xmlNode.) The function argument really shouldn't be part of the template pack, only `std::function<T(const char*)>` will work.

The only differences for the second function are the return type (arbitrary in the first, gboolean in the second) and that the function argument is `std::function<gboolean(Obj*, const char*)>`. Using the different function argument also moves you from template type deduction to normal function argument type resolution so you can use the same name for both. I think this would be a good place to do that.

christopherlam commented:
I think this is nearly complete... The goal was to replicate all `dom_tree_to_text` with new templated functions. Hopefully more eyes == less bugs. My cli to load xml load and run report improves from 5.0s to 4.0s.

christopherlam commented:
On reflection the helper functions should have their arguments swapped around; should be the following to be more "intuitive"/"natural"/etc. Will try to modify.

`apply_xmlnode_text(f, node, [default])` --> returns `f(node_text);` or `default`
`apply_xmlnode_text(f, obj, node)` --> calls `f(obj, node_text);` returning `true` if ok

similar to scheme `(apply f (list arg1 arg2...))` is equivalent to `(f arg1 arg2...)`

jralls commented:
> On reflection the helper functions should have their arguments swapped around; should be the following to be more "intuitive"/"natural"/etc.

Makes sense.

jralls commented:
But you're still passing `func` as an unconstrained template parameter. That works but it will produce hard to understand errors if somebody tries to pass in something that doesn't match the signature of how it's used.

christopherlam commented:
added thanks to chatgpt...

because std::function has overhead
and function pointer cannot take lambda, and cannot inline
and F&& with static_assert is zero-cost

jralls commented:
I think chatgpt gave you the wrong answer, or maybe you asked the wrong question.  It's far better for the compiler to say that it can't find a matching `apply_xmlmode_text` at the line with the wrong parameters than to raise an assert fail inside `apply_xmlnode_text` with the source of the problem buried in a note message somewhere after the error.

It also doesn't make sense to pretend the function is generic in a way that it isn't: F is exactly one type, `std::function` in each version of `apply_xmlmode_text`. Tell the compiler that up front. The simplest way to do that is to specify the type in the function arguments without templating. There's a more complicated way using `std::enable_if` on the template paramter, but the extra work and clunkiness doesn't get you anything. (C++20 adds "concepts light" that provides a prettier way to constrain the template type deduction but it's still extra work that doesn't make sense when there's only one type that satisfies the template.
```
template<typename T>
inline T
apply_xmlnode_text (std::function<T, xmlNodePtr, T> f, xmlNodePtr node, T default_val = T{})

template<typename Obj>
inline bool
apply_xmlnode_text (std::function<void, Obj*, const char*> f, Obj* obj, xmlNodePtr node)
```

BTW there's a stray `;` at the end of the first definition.

christopherlam commented:
it should be `std::function<T(const char*)>` but std::function has runtime overhead, and I can't currently make it compile.

jralls commented:
```
Template<typename T, typename F, typename std::enable_if<std::is_invocable_r_v<void, F&, Obj*, const char*>, int> == 0>
inline T
apply_xmlnode_text (F&& f, xmlNodePtr node, T default_val = T{})
```
Is the ugly way. It saves the virtual lookup from std::function and won't break inlining.

What compile error are you getting from std::function?

christopherlam commented:
> ```
> Template<typename T, typename F, typename std::enable_if<std::is_invocable_r_v<void, F&, Obj*, const char*>, int> == 0>
> inline T
> apply_xmlnode_text (F&& f, xmlNodePtr node, T default_val = T{})
> ```
> 
> Is the ugly way. It saves the virtual lookup from std::function and won't break inlining.
> 
> What compile error are you getting from std::function?

```
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/gnc-address-xml-v2.cpp: In function ‘gboolean address_name_handler(xmlNodePtr, gpointer)’:
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/gnc-address-xml-v2.cpp:105:31: error: no matching function for call to ‘apply_xmlnode_text(void (&)(GncAddress*, const char*), GncAddress*&, xmlNode*&)’
  105 |     return apply_xmlnode_text (gncAddressSetName, pdata->address, node);
      |            ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/chris/sources/gnucash/stable/libgnucash/backend/xml/gnc-address-xml-v2.cpp:36:
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/sixtp-dom-parsers.h:79:1: note: candidate: ‘template<class T> T apply_xmlnode_text(std::function<T(const char*)>, xmlNodePtr, T)’
   79 | apply_xmlnode_text (std::function<T(const char*)> f, xmlNodePtr node, T default_val = T{})
      | ^~~~~~~~~~~~~~~~~~
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/sixtp-dom-parsers.h:79:1: note:   template argument deduction/substitution failed:
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/gnc-address-xml-v2.cpp:105:31: note:   mismatched types ‘std::function<T(const char*)>’ and ‘void (*)(GncAddress*, const char*)’ {aka ‘void (*)(_gncAddress*, const char*)’}
  105 |     return apply_xmlnode_text (gncAddressSetName, pdata->address, node);
      |            ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/sixtp-dom-parsers.h:95:1: note: candidate: ‘template<class Obj> bool apply_xmlnode_text(std::function<void(Obj*, const char*)>, Obj*, xmlNodePtr)’
   95 | apply_xmlnode_text (std::function<void(Obj*, const char*)> f, Obj* obj, xmlNodePtr node)
      | ^~~~~~~~~~~~~~~~~~
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/sixtp-dom-parsers.h:95:1: note:   template argument deduction/substitution failed:
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/gnc-address-xml-v2.cpp:105:31: note:   mismatched types ‘std::function<void(Obj*, const char*)>’ and ‘void (*)(GncAddress*, const char*)’ {aka ‘void (*)(_gncAddress*, const char*)’}
  105 |     return apply_xmlnode_text (gncAddressSetName, pdata->address, node);
      |            ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/gnc-address-xml-v2.cpp: In function ‘gboolean address_addr1_handler(xmlNodePtr, gpointer)’:
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/gnc-address-xml-v2.cpp:113:31: error: no matching function for call to ‘apply_xmlnode_text(void (&)(GncAddress*, const char*), GncAddress*&, xmlNode*&)’
  113 |     return apply_xmlnode_text (gncAddressSetAddr1, pdata->address, node);
      |            ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/chris/sources/gnucash/stable/libgnucash/backend/xml/sixtp-dom-parsers.h:79:1: note: candidate: ‘template<class T> T apply_xmlnode_text(std::function<T(const char*)>, xmlNodePtr, T)’
   79 | apply_xmlnode_text (std::function<T(const char*)> f, xmlNodePtr node, T default_val = T{})
      | ^~~~~~~~~~~~~~~~~~
```

jralls commented:
Apparently it's too hard for compilers (clang has the same problem) to both convert a C-style function ptr to std::function and perform template type deduction on the template arguments to std::function in one call. You can work around it by e.g.
```apply_xmlnode_text(static_cast<std::function<void(GncAddress*, const char*)>(gncAddressSetName), pdata->address, node);``` but that's just as ugly and in more places than enable_if on the templates.

I had a couple of errors in the template decls, so here are the tested-correct ones:
```
template <typename T, typename F,
          std::enable_if_t<std::is_invocable_r_v<void, F, const char*>, int> = 0>
inline T
apply_xmlnode_text (F&& f, xmlNodePtr node, T default_val = T{})

template <typename Obj, typename F,
    std::enable_if_t<std::is_invocable_r_v<void, F, Obj*, const char*>, int> = 0>
inline bool
apply_xmlnode_text (F&& f, Obj* obj, xmlNodePtr node)
```

To illustrate the difference in the error messages I mutated the lambda in the second `apply_xmlnode_text` to `    auto set_str = [&f](Obj* obj, auto txt)`. With enable_if the error is
```
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:110:12: error: no matching function for call to 'apply_xmlnode_text'
  110 |     return apply_xmlnode_text<bool> (set_str, node, false);
      |            ^~~~~~~~~~~~~~~~~~~~~~~~
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/gnc-account-xml-v2.cpp:179:12: note: in instantiation of function template specialization 'apply_xmlnode_text<account_s, void (&)(account_s *, const char *), 0>' requested here
  179 |     return apply_xmlnode_text (xaccAccountSetName, pdata->account, node);
      |            ^
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:80:1: note: candidate template ignored: requirement 'std::is_invocable_r_v<void, (lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:105:20) &, const char *>' was not satisfied [with T = bool, F = (lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:105:20) &]
   80 | apply_xmlnode_text (F&& f, xmlNodePtr node, T default_val = T{})
      | ^
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:100:1: note: candidate template ignored: requirement 'std::is_invocable_r_v<void, (lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:105:20) &, bool *, const char *>' was not satisfied [with Obj = bool, F = (lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:105:20) &]
  100 | apply_xmlnode_text (F&& f, Obj* obj, xmlNodePtr node)
      | ^
1 error generated.
```
and with static_assert it's
```
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:81:20: error: static assertion failed due to requirement 'std::is_invocable_r_v<bool, (lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:104:20) &, const char *>': apply_xmlnode_text expects a callable F(const char*) -> T
   81 |     static_assert (std::is_invocable_r_v<T, F&, const char*>,
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:109:12: note: in instantiation of function template specialization 'apply_xmlnode_text<bool, (lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:104:20) &>' requested here
  109 |     return apply_xmlnode_text<bool> (set_str, node, false);
      |            ^
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/gnc-account-xml-v2.cpp:179:12: note: in instantiation of function template specialization 'apply_xmlnode_text<account_s, void (&)(account_s *, const char *), 0>' requested here
  179 |     return apply_xmlnode_text (xaccAccountSetName, pdata->account, node);
      |            ^
In file included from /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/gnc-account-xml-v2.cpp:38:
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:88:16: error: no matching function for call to object of type '(lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:104:20)'
   88 |         return f(txt);
      |                ^
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:104:20: note: candidate function template not viable: requires 2 arguments, but 1 was provided
  104 |     auto set_str = [&f](Obj* obj, auto txt)
      |                    ^    ~~~~~~~~~~~~~~~~~~
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:91:16: error: no matching function for call to object of type '(lambda at /Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:104:20)'
   91 |         return f(txt->c_str());
      |                ^
/Users/john/Development/Gnucash-git-11.0-arm64/src/gnucash-git/libgnucash/backend/xml/sixtp-dom-parsers.h:104:20: note: candidate function template not viable: requires 2 arguments, but 1 was provided
  104 |     auto set_str = [&f](Obj* obj, auto txt)
      |                    ^    ~~~~~~~~~~~~~~~~~~
3 errors generated.
```

christopherlam commented:
thanks this did it...

jralls reviewed (approved):
OK, merge when ready.

jralls commented:
@christopherlam You don't have to rebase before merging unless there are conflicts that you don't want to resolve in the merge commit.

christopherlam commented:
ugh. I think we need to revert `apply_xmlnode_text` and `[backend/xml] dom_tree_to_text returns std::optional<std::string>`. there's leaks. can't figure it out.
[Valgrind memory leak analysis.html](https://github.com/user-attachments/files/24157336/Valgrind.memory.leak.analysis.html)

jralls commented:
Most of what ChatGPT is complaining about is either nonsensical (this isn't multi-threaded code so `apply_xmlnode_text` can't return early. There's no path where it can call what ChatGPT is calling a handler more than once.)
It complains
```
f (auto txt = dom_node_to_text(node))
    ...
else
    txt = dom_tree_to_text(node);

This means:
   *  borrowed vs owned strings
   *  different lifetimes
   *  different failure behaviour
```
That ignores that in the second case we're passing txt->c_str(), i.e. a borrowed string, to `f()` and that in all cases `f()` immediately copies the `char*` and then the std::string if there is one goes out of scope so there's no lifetime issue.

Different failure behavior is a valid complaint and might be the cause of some of the leaks.
For example the lambda in `owner_type_handler` returns `true` even if the type doesn't match one of the choices; It doesn't know that that's defensive and that it's not actually possible. Nevertheless it would be reasonable to have the lambda throw a `std::runtime_error` if it falls through. Looks like a similar problem in `dom_tree_to_guid` and perhaps others. 

I dunno how you constructed valgrind_memory_leak_analysis.html but the valgrind output is rendered illegible for lack of being wrapped in `<pre>...</pre>`: HTML converts all whitespace including newlines to a single space so the output is rendered as a single line. I suggest you go through the changes and restore all of the error conditions to the lambdas and see if that fixes the leaks. The exceptions will make the resulting object code a little bigger and maybe a little slower. If it turns out to be a lot slower then we can consider another error handling mechanism: Maybe the lambdas could capture a bool& success and set it to true if they succeed. Also any logging in the original should also be in the lambdas.

christopherlam commented:
Maybe best revert for release, I'll do tomorrow.

jralls commented:
How about just reverting the replacements where the lambda doesn't do what the original code did?

christopherlam commented:
> How about just reverting the replacements where the lambda doesn't do what the original code did?

Uh because I'm not confident I can cleanly identify the offending functions, and thought it'd be cleaner to go back to the drawing board

jralls commented:
I've taken a [stab at a partial revert](https://github.com/jralls/gnucash/tree/revert-xmlnode) but it still needs some cleanup to build. I'll finish that up tomorrow morning and you can try it in Valgrind.

Was yesterday's Valgrind run the test suite or did you open GnuCash and try some stuff?

christopherlam commented:
> I've taken a [stab at a partial revert](https://github.com/jralls/gnucash/tree/revert-xmlnode) but it still needs some cleanup to build. I'll finish that up tomorrow morning and you can try it in Valgrind.
> 
> Was yesterday's Valgrind run the test suite or did you open GnuCash and try some stuff?

It's from a regular datafile.

christopherlam commented:
ok as of current `stable` theres some valgrind logs

```

==465037== 768 (464 direct, 304 indirect) bytes in 2 blocks are definitely lost in loss record 21,595 of 22,755
==465037==    at 0x5C5DCCD: g_type_create_instance (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x5C4222F: ??? (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x5C434DD: g_object_new_with_properties (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x5C44310: g_object_new (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x62BC4C2: gncInvoiceCreate (gncInvoice.c:319)
==465037==    by 0x5DD65D4: entry_invoice_handler(_xmlNode*, void*) (gnc-entry-xml-v2.cpp:554)
==465037==    by 0x5DEB266: gnc_xml_set_data (sixtp-dom-parsers.cpp:694)
==465037==    by 0x5DEB266: dom_tree_generic_parse(_xmlNode*, dom_tree_handler*, void*) (sixtp-dom-parsers.cpp:725)
==465037==    by 0x5DD625C: dom_tree_to_entry (gnc-entry-xml-v2.cpp:684)
==465037==    by 0x5DD625C: gnc_entry_end_handler(void*, _GSList*, _GSList*, void*, void*, void**, char const*) (gnc-entry-xml-v2.cpp:732)
==465037==    by 0x5DEFA67: sixtp_sax_end_handler(void*, unsigned char const*) (sixtp.cpp:540)
==465037==    by 0x5F37BC7: ??? (in /gnu/store/71v7zvadm07b7zh1z22d9242xln0q843-libxml2-2.14.6/lib/libxml2.so.16.0.6)
==465037==    by 0x5F384B6: ??? (in /gnu/store/71v7zvadm07b7zh1z22d9242xln0q843-libxml2-2.14.6/lib/libxml2.so.16.0.6)
==465037==    by 0x5F3E137: ??? (in /gnu/store/71v7zvadm07b7zh1z22d9242xln0q843-libxml2-2.14.6/lib/libxml2.so.16.0.6)
==465037==


==465037== 288 (224 direct, 64 indirect) bytes in 4 blocks are definitely lost in loss record 20,876 of 22,755
==465037==    at 0x48457B2: malloc (in /gnu/store/3i5px8576v3zw9836cz5ji1zk1ffbyki-valgrind-3.26.0/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==465037==    by 0x693C27D: g_malloc (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libglib-2.0.so.0.8303.0)
==465037==    by 0x6912819: ??? (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libglib-2.0.so.0.8303.0)
==465037==    by 0x6912E99: ??? (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libglib-2.0.so.0.8303.0)
==465037==    by 0x5C40994: ??? (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x5C41017: g_object_unref (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x62D7C90: qof_commit_edit_part2 (qofinstance.cpp:1035)
==465037==    by 0x5DDD11F: taxtable_guid_handler(_xmlNode*, void*) (gnc-tax-table-xml-v2.cpp:261)
==465037==    by 0x5DEB266: gnc_xml_set_data (sixtp-dom-parsers.cpp:694)
==465037==    by 0x5DEB266: dom_tree_generic_parse(_xmlNode*, dom_tree_handler*, void*) (sixtp-dom-parsers.cpp:725)
==465037==    by 0x5DDCB23: dom_tree_to_taxtable (gnc-tax-table-xml-v2.cpp:378)
==465037==    by 0x5DDCB23: gnc_taxtable_end_handler(void*, _GSList*, _GSList*, void*, void*, void**, char const*) (gnc-tax-table-xml-v2.cpp:418)
==465037==    by 0x5DEFA67: sixtp_sax_end_handler(void*, unsigned char const*) (sixtp.cpp:540)
==465037==    by 0x5F37BC7: ??? (in /gnu/store/71v7zvadm07b7zh1z22d9242xln0q843-libxml2-2.14.6/lib/libxml2.so.16.0.6)
==465037== 


==465037== 48 bytes in 1 blocks are definitely lost in loss record 13,283 of 22,755
==465037==    at 0x4845F18: operator new(unsigned long) (in /gnu/store/3i5px8576v3zw9836cz5ji1zk1ffbyki-valgrind-3.26.0/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==465037==    by 0x62D5C34: qof_instance_init(QofInstance_s*) (qofinstance.cpp:262)
==465037==    by 0x5C5DBD2: g_type_create_instance (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x5C4222F: ??? (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x5C434DD: g_object_new_with_properties (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x5C44310: g_object_new (in /gnu/store/r99aqxzrfkhcijp2ni72nzcahynb9kkv-glib-2.83.3/lib/libgobject-2.0.so.0.8303.0)
==465037==    by 0x62C4EE2: gncTaxTableCreate (gncTaxTable.c:411)
==465037==    by 0x5DDCFBA: set_parent_child(_xmlNode*, taxtable_pdata*, void (*)(_gncTaxTable*, _gncTaxTable*)) (gnc-tax-table-xml-v2.cpp:239)
==465037==    by 0x5DEB266: gnc_xml_set_data (sixtp-dom-parsers.cpp:694)
==465037==    by 0x5DEB266: dom_tree_generic_parse(_xmlNode*, dom_tree_handler*, void*) (sixtp-dom-parsers.cpp:725)
==465037==    by 0x5DDCB23: dom_tree_to_taxtable (gnc-tax-table-xml-v2.cpp:378)
==465037==    by 0x5DDCB23: gnc_taxtable_end_handler(void*, _GSList*, _GSList*, void*, void*, void**, char const*) (gnc-tax-table-xml-v2.cpp:418)
==465037==    by 0x5DEFA67: sixtp_sax_end_handler(void*, unsigned char const*) (sixtp.cpp:540)
==465037==    by 0x5F37BC7: ??? (in /gnu/store/71v7zvadm07b7zh1z22d9242xln0q843-libxml2-2.14.6/lib/libxml2.so.16.0.6)
==465037== 
```

and I missed out on `g_date_free` in `dom_tree_to_gdate_kvp_value`:

```diff
modified   libgnucash/backend/xml/sixtp-dom-parsers.cpp
@@ -176,7 +176,10 @@ static KvpValue*
 dom_tree_to_gdate_kvp_value (xmlNodePtr node)
 {
     auto date = dom_tree_to_gdate (node);
-    return date ? new KvpValue {*date} : nullptr;
+    if (!date) return nullptr;
+    auto rv = new KvpValue {*date};
+    g_date_free (date);
+    return rv;
 }
```

jralls commented:
The [partial revert\(https://github.com/jralls/gnucash/tree/revert-xmlnode) is cleaned up so you can try valgrind on it. 

Are those 3 the only leaks in stable?

jralls commented:
Are you sure that those leaks are new? None of them look to me like they have anything to do with this refactor.

jralls commented:
I think this is the cause of #3:
```
static void
gnc_taxtable_finalize(GObject* ttp)
{
    G_OBJECT_CLASS(gnc_taxtable_parent_class)->dispose(ttp);
}
```
That should be *finalize*, not *dispose*.
